### ISSUE TITLE
Throw in closure triggers false positive DocBlockThrows.ThrowMissing

### ISSUE THREAD

**Issue by levacic** [spryker/code-sniffer#238]
```php
<?php

declare(strict_types=1);

namespace App;

use Closure;
use RuntimeException;

class InvalidThrowSniff
{
    /**
     * Return a closure that throws a runtime exception.
     *
     * @return Closure
     */
    public function foo(): Closure
    {
        return static function (): void {
            throw new RuntimeException();
        };
    }
}
```

This code triggers a false positive `Spryker.Commenting.DocBlockThrows.ThrowMissing` for the `foo()` method's docblock.

It's a false positive because the `foo()` method never throws.

**Comment by dereuromark** [comment]
Do you have an idea for a fix? We probably have to check scope or depth. A PR would be wonderful .

**Comment by levacic** [comment]
I could give it a shot but unfortunately I have zero experience writing sniffs so I'd need some help.

A quick guess would be that [this loop](https://github.com/spryker/code-sniffer/blob/9cacaae777d21391f531755677997506d378bd8a/Spryker/Sniffs/Commenting/DocBlockThrowsSniff.php#L101) should keep track of the current scope depth. Here's some pseudo-code:

- before the loop, do `$currentScopeDepth = 0;`
- at the beginning of the loop, before comparing the current token to `T_THROW`, check if the current token is opening a new scope; if it is, increment `$currentScopeDepth` and `continue`
- the next check should be whether the current token is closing a scope - if so, decrement `$currentScopeDepth` and `continue`
- the next check should be whether `$currentScopeDepth > 0` (and `continue` if it is, thus ignoring nested scopes)
- after that, the rest of the code as it is should remain the same (ie. checking whether the current token is `TO_THROW`)

It seems like it would be necessary to track the scope depth like this because we might have nested scopes, and we might go back and forth between scope depth within a single method (e.g. a method that defines several closures).

That's just off the top of my head, but it's late here, it's been a long day, and I might have missed something. If this does make sense, however, and if you could let me know how to correctly check whether the current token is opening or closing a new scope, I guess I could give a shot at a PR.

**Comment by dereuromark** [comment]
Feel free to give it a try.

**Comment by levacic** [comment]
Eh, just realized it's not as simple - `T_OPEN_CURLY_BRACKET` can be used in other context that don't create a new scope (e.g. `if` blocks).

Instead of `$currentScopeDepth` it seems like we might need some stack-based tracking that also stores `T_CLOSURE` tokens, and possibly others - what else can result in a nested scope?

I guess something like this should also be considered:

```php
/**
 * Return a closure that throws a runtime exception when passed `true`.
 *
 * @return Closure
 */
public function foo()
{
    return fn($value) => $value === true ? throw new RuntimeException() : null;
}
```

Granted this seems like a weird construct but it does start a new scope - this time without brackets at all.

So we would need to also track `T_FN` as well - which could be more complex than this example as well.

Any ideas how to proceed on this?

P.S. It seems that the test suite is currently failing (and only contains a couple of tests anyway). Is that correct, or have I missed something?

**Comment by dereuromark** [comment]
I do have a follow topic I found, that also doesn't get work properly:

```
    /**
     * @param array $data
     * @param int $jobId
     *
     * @throws \RuntimeException
     *
     * @return void
     */
    public function run(array $data, int $jobId): void
    {
        $repository = $data['repository'];
        $branch = $data['branch'];

        [$organization, $repository] = explode('/', $repository, 2);
        $processor = new ProjectProcessor($organization, $repository);
        $processor->checkout($branch);

        $processor->updateDockerSubmodule($branch);

        try {
            $processor->composerUpdate($branch, false, true);
        } catch (RuntimeException $e) {
            if (strpos($e->getMessage(), 'does not contain valid JSON') === false) {
                throw $e;
            }

            $processor->composerUpdateFallback($branch);
        }
```

Produces

> | ERROR | Throw token found, but no annotation for it.
    |       | (Spryker.Commenting.DocBlockThrows.ThrowAnnotationMissing)


We can address this in a separate topic/PR.
